### PR TITLE
feat: add RA.move

### LINKED ISSUE #485: RA.move
Get a new list with an item's index changed.

```
const moveItem = curry(
  (oldIdx, newIdx, list) => compose(
      insert(
        newIdx,
        nth(oldIdx, list)
      ),
      remove(
        oldIdx,
        1
      )
    )(list)
)

const a = [1, 2, 3, 4] 

moveItem(0, 3, a) // => [2, 3, 4, 1]
```

[REPL](http://ramdajs.com/repl/#?const%20moveItem%20%3D%20curry%28%0A%20%20%28oldIdx%2C%20newIdx%2C%20list%29%20%3D%3E%20compose%28%0A%20%20%20%20%20%20insert%28%0A%20%20%20%20%20%20%20%20newIdx%2C%0A%20%20%20%20%20%20%20%20nth%28oldIdx%2C%20list%29%0A%20%20%20%20%20%20%29%2C%0A%20%20%20%20%20%20remove%28%0A%20%20%20%20%20%20%20%20oldIdx%2C%0A%20%20%20%20%20%20%20%201%0A%20%20%20%20%20%20%29%0A%20%20%20%20%29%28list%29%0A%29%0A%0Aconst%20a%20%3D%20%5B1%2C%202%2C%203%2C%204%5D%20%0A%0AmoveItem%280%2C%203%2C%20a%29%20%2F%2F%20%3D%3E%20%5B2%2C%203%2C%204%2C%201%5D%0AmoveItem%283%2C%200%2C%20a%29%20%2F%2F%20%3D%3E%20%5B4%2C%201%2C%202%2C%203%5D%0AmoveItem%282%2C%203%2C%20a%29%20%2F%2F%20%3D%3E%20%5B1%2C%202%20%2C4%20%2C3%5D%0AmoveItem%281%2C%200%2C%20a%29%20%2F%2F%20%3D%3E%20%5B2%2C%201%20%2C3%20%2C4%5D%0AmoveItem%280%2C%201%2C%20a%29%20%2F%2F%20%3D%3E%20%5B2%2C%201%20%2C3%20%2C4%5D%0A)

Additional possibilities:

- `RA.moveBy(fromIndex, by)`

### PR DESCRIPTION
**Pull request by Undistraction**
Adds `RA.move`

> Returns a new list with the item at the position `fromIdx` moved to the position `toIdx`. If the `toIdx` is greater than the number of items - 1, it will be placed in the last position.

In its present form it uses `R.insert` and `R.remove` both of which have undefined behaviour if negative indexes are used, so `RA.move` doesn't support negative indexes either. Is this something we should support?

Closes #485

### CODE REVIEW

File: test/move.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  });
+
+  it('should support currying', function() {
+    const expected = ['a', 'c', 'd', 'b', 'e'];
+    eq(RA.move(1)(3, list), expected);
+    eq(RA.move(1, 3)(list), expected);
+    eq(RA.move(1)(3)(list), expected);
+  });
+
+  it('should treat nested lists correctly', function() {
```
Comment (guillaumearm) [inline test/move.js]: IMHO, this test has no sense, this works on numbers, objects or anything else, it's the same thing.
The problem here is that test suggests you can move some elements from nested list to another.

Comment (Undistraction) [reply test/move.js]: I just wanted to ensure that there wasn't any unpacking/flattening of nested lists. I'm happy to remove it though.

Comment (guillaumearm) [reply test/move.js]: Nice, so I think you can just rename it ;-)

File: src/index.d.ts [hunk 2]
```diff
@@ -933,6 +933,18 @@ declare namespace RamdaAdjunct {
          */
         dispatch(functions: Function[]): Function;
 
+        /**
+         * Returns a new list with the item at the position `fromIdx` moved to the position `toIdx`.
+         * If the * `toIdx` is greater than the number of items - 1, it will be placed in the last
```
Comment (char0n) [inline src/index.d.ts]:
```js
If the * `toIdx`
```
What does `*` represents here ?

Comment (char0n) [reply src/index.d.ts]: I think you can simplify and generalize second sentence  into 
```
If the `toIdx` is out of the `list` range, the item will be placed on the last position in the `list`.
If the `fromIdx` is out the `list` range, ...what happens here... ?
```

File: src/move.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * @sig Number -> Number -> [a] -> [a]
+ * @param {Number} fromIdx Position of item to be moved
+ * @param {Number} toIdx Position of item after move
+ * @return {Array}
+ * @example
+ *
+ * const list = ['a', 'b', 'c', 'd', 'e'];
+ * RA.move(1, 3, list) //=> ['a', 'c', 'd', 'b', 'e']
+ */
+const moveItem = curry((oldIdx, newIdx, list) =>
```
Comment (char0n) [inline src/move.js]:
```js
const moveItem -> const move
```
Or is there some other reason why this is called `moveItem` ?

File: test/move.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  it('should move an item to a higher idx', function() {
+    const expected = ['a', 'c', 'd', 'b', 'e'];
+    eq(RA.move(1, 3, list), expected);
+  });
+
+  it('should move an item to a lower idx', function() {
+    const expected = ['a', 'd', 'b', 'c', 'e'];
+    eq(RA.move(3, 1, list), expected);
+  });
+
```
Comment (char0n) [inline test/move.js]: We're missing test for a case when `fromIdx` is out the list range.

Comment (char0n) [reply test/move.js]: Well I think we should support the same as the `Array` supports.
```javascript
const list = ['a', 'b', 'c'];
list[-1]  = 'd'; list.length = 3;
list[100] = 'd'; list.length = 101;
```
...and so on. Does it make sense ?

Comment (Undistraction) [reply test/move.js]: > Now I have noticed that ramda doesn't support negative indexes and the behaviour is undefined when they are supplied

Exactly. 

> we should only add test describing what happens if fromIdx is out the list range (undefined added to toIndex)

I'll add this now.

Comment (Undistraction) [reply test/move.js]: @char0n That should be good to merge.

Comment (char0n) [reply test/move.js]: Will do the final check tommorow

File: src/move.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import { curry, compose, insert, nth, remove } from 'ramda';
+
+/**
+ * Returns a new list with the item at the position `fromIdx` moved to the position `toIdx`. If the
+ * `toIdx` is out of the `list` range, the item will be placed at the last position of the `list`.
+ * Negative indices are not supported.
+ *
+ * @func move
+ * @memberOf RA
+ * @since {@link https://char0n.github.io/ramda-adjunct/2.7.0|v2.7.0}
```
Comment (char0n) [inline src/move.js]: 2.8.0

File: src/index.d.ts [hunk 6]
```diff
@@ -949,6 +948,18 @@ declare namespace RamdaAdjunct {
          */
         dispatch(functions: Function[]): Function;
 
+        /**
+         * Returns a new list with the item at the position `fromIdx` moved to the position `toIdx`.
+         * If the `toIdx` is out of the `list` range, the item will be placed at the last position
+         * of the `list`. Nagative indices are not supported.
```
Comment (char0n) [inline src/index.d.ts]: Would rephrase the `Nagative indices are not supported` to `When negative indices are provided the behavior of the move is unspecified`.

**Comment by char0n** [comment]
@Undistraction can we resolve conflicts here and rebase to master ? Nothing is preventing this for getting into next release.

**Comment by Undistraction** [comment]
@char0n Are you happy for me to squash and merge?

**Comment by char0n** [comment]
@Undistraction yep go for it!
